core/desugarDatatypes.ml

```diff
@@ -41,6 +41,71 @@ let concrete_subkind =
   | Some subkind -> subkind
   | None         -> default_subkind
 
+(** Replaces all placeholder variables with a fresh one. *)
+class basic_freshener = object
+  inherit SugarTraversals.map as super
+
+  method! known_type_variable =
+    let module SC = SugarConstructors.SugartypesPositions in
+    function
+    | (("_" | "_anon"), None, freedom) ->
+       SC.fresh_known_type_variable freedom
+       |> super#known_type_variable
+    | v ->  super#known_type_variable v
+end
+
+let freshen_vars =
+  let module SC = SugarConstructors.SugartypesPositions in
+  (* Determine if this is an "anonymous" effect type variable, and so introduced
+     within a simple arrow (->, ~>, -@, ~>) *)
+  let is_anon_effect = function
+    | Datatype.Open ("_anon", None, `Rigid) -> true
+    | _ -> false in
+  (* If this function type is exclusively composed of anonymous effect type
+     variables. Or rather, there are no explicitly mentioned effect variables. *)
+  let all_anon_effects = object
+    inherit SugarTraversals.fold as super
+
+    val all_anon = true
+    method all_anon = all_anon
+
+    method! datatypenode = let open Datatype in
+      function
+      | Function (_, (_, eff_var), _) | Lolli (_, (_, eff_var), _)
+           when not (is_anon_effect eff_var) -> {<all_anon = false>}
+      | ty -> super#datatypenode ty
+  end in
+
+  let basic_refresh = new basic_freshener
+  and shared_refresh var = object
+    inherit basic_freshener as super
+    method! datatypenode = let open Datatype in
+      function
+      | Function (args, (effects, eff_var), ret) when is_anon_effect eff_var
+        -> super#datatypenode (Function (args, (effects, var), ret))
+      | Lolli (args, (effects, eff_var), ret) when is_anon_effect eff_var
+        -> super#datatypenode (Lolli (args, (effects, var), ret))
+      | ty -> super#datatypenode ty
+  end in
+
+  object
+    inherit basic_freshener as super
+    method! datatypenode = let open Datatype in
+      function
+      | (Function _ | Lolli _) as dt ->
+         (* If shared_effect_vars is enabled, and all variables are fresh (and so
+            not explicitly named), then we will substitute the same variable
+            across all arrows. Otherwise every arrow gets its own row
+            variable, as normal. *)
+         if Settings.get_value Basicsettings.Types.shared_effect_vars
+            && (all_anon_effects#datatypenode dt)#all_anon then
```

Comment:
One way to make this a linear time algorithm is to use non-linear control and apply the syntactic sugar optimistically. Use three objects: one to traverse the term structure, let's call simply call it `effect_var_desugarer`, and two objects for refreshing effect type variables akin to your `shared_refresh` and `basic_refresh`.

Whenever the `effect_var_desugarer` comes across a `datatypenode` it has to pick between the two visitors. It should optimistically select the `shared_refresh` visitor. The `shared_refresh` visitor should optimistically rewrite a given datatype, however, if it comes across an effect variable that violates the assumption, then it should throw an exception. This exception should be caught by the `effect_var_desugarer`, causing it to apply the `basic_refresh` visitor instead.

This algorithm should have linear time complexity. Let `n` be the size of the datatype, in the worst case we visit the whole datatype before discovering that there is a variable which violates the assumption. Moreover, the depth of call stack relative to the initial invocation of the visitor can at most be `n`. So it takes `2*n` time to discover and propagate the information back to the `effect_var_desugarer` visitor, which then applies the other visitor, which must necessarily visit the entire datatype and return again, requiring at most `2*n` time also. Summing up `2*n + 2*n = O(n)`.